sdks/python/apache_beam/runners/direct/executor.py

```diff
@@ -291,45 +294,58 @@ def call(self):
 
     side_input_values = [self._side_input_values[side_input]
                          for side_input in self._applied_ptransform.side_inputs]
-
     try:
```

Comment:
<!--new_thread; commit:eb2acfa855db99ab4b826344656ec29dcdb175e3; resolved:0-->
Can you track the retry_count here (removing it from `self._retry_count`) and use a while loop (`while retry_count < 4`) here to track retries, instead of using `@retry`?  The way you are using `retrying.retry` here makes it hard to do things like report an intermediate failure, as we should do on line 301.